File: src/coordinate.rs
use std::cmp::Ordering;

use bitflags::bitflags;

bitflags! {
    pub struct Direction: u8 {
        const LEFT  = 0x1;
        const RIGHT = 0x2;
        const UP    = 0x4;
        const DOWN  = 0x8;
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Coordinate {
    pub x: usize,
    pub y: usize,
}

impl Coordinate {
    #[inline]
    pub const fn new(x: usize, y: usize) -> Self {
        Self { x, y }
    }

    pub fn raw_adjacent(self, direction: Direction) -> Self {
        let mut output = self;

        if direction & Direction::LEFT == Direction::LEFT {
            output.x -= 1;
        }
        if direction & Direction::RIGHT == Direction::RIGHT {
            output.x += 1;
        }
        if direction & Direction::UP == Direction::UP {
            output.y += 1;
        }
        if direction & Direction::DOWN == Direction::DOWN {
            output.y -= 1;
        }

        output
    }

    pub fn adjacent(self, direction: Direction) -> Option<Self> {
        let mut output = self;

        if direction & Direction::LEFT == Direction::LEFT {
            if output.x == 0 {
                return None;
            }
            output.x -= 1;
        }
        if direction & Direction::RIGHT == Direction::RIGHT {
            output.x += 1;
        }
        if direction & Direction::UP == Direction::UP {
            output.y += 1;
        }
        if direction & Direction::DOWN == Direction::DOWN {
            if output.y == 0 {
                return None;
            }
            output.y -= 1;
        }

        Some(output)
    }
}

impl PartialOrd for Coordinate {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(if self.y == other.y {
            self.x.cmp(&other.x)
        } else {
            self.y.cmp(&other.y)
        })
    }
}

pub struct CoordinateIter {
    current: Coordinate,
    end: Coordinate,
    row_limit: usize,
}

impl CoordinateIter {
    pub fn new(start: Coordinate, end: Coordinate, row_limit: usize) -> Self {
        Self {
            current: start,
            end,
            row_limit,
        }
    }
}

impl Iterator for CoordinateIter {
    type Item = Coordinate;
    fn next(&mut self) -> Option<Self::Item> {
        if self.current > self.end {
            return None;
        }

        let output = self.current;

        if self.current.x >= self.row_limit {
            self.current = Coordinate::new(0, self.current.y + 1);
        } else {
            self.current = Coordinate::new(self.current.x + 1, self.current.y);
        }

        Some(output)
    }
}

File: src/Coordinate.zig
// Coordinate fields
x: usize,
y: usize,
// END fields

pub const Coordinate = @This();
const Self = @This();

pub fn new(x: usize, y: usize) Self {
    return .{ .x = x, .y = y };
}

pub fn raw_adjacent(
    self: Self,
    comptime left: bool,
    comptime right: bool,
    comptime up: bool,
    comptime down: bool,
) Self {
    var output = self;

    if (left) {
        output.x -= 1;
    }
    if (right) {
        output.x += 1;
    }
    if (up) {
        output.y += 1;
    }
    if (down) {
        output.y -= 1;
    }

    return output;
}

pub fn adjacent(
    self: Self,
    comptime left: bool,
    comptime right: bool,
    comptime up: bool,
    comptime down: bool,
) ?Self {
    var output = self;

    if (left) {
        if (output.x == 0) {
            return null;
        }
        output.x -= 1;
    }
    if (right) {
        output.x += 1;
    }
    if (up) {
        output.y += 1;
    }
    if (down) {
        if (output.y == 0) {
            return null;
        }
        output.y -= 1;
    }

    return output;
}

pub const Compare = enum {
    eq,
    neq,
    less,
    less_eq,
    great,
    great_eq,
};

pub fn cmp(self: Self, comptime compare: Compare, rhs: Self) bool {
    switch (compare) {
        .eq => return self.x == rhs.x and self.y == rhs.y,
        .neq => return !self.cmp(.eq, rhs),
        .less => {
            if (self.y == rhs.y) {
                return self.x < rhs.x;
            } else {
                return self.y < rhs.y;
            }
        },
        .less_eq => {
            if (self.y == rhs.y) {
                return self.x <= rhs.x;
            } else {
                return self.y <= rhs.y;
            }
        },
        .great => return !self.cmp(.less_eq, rhs),
        .great_eq => return !self.cmp(.less, rhs),
    }
}

pub const CoordinateIterator = struct {
    current: Coordinate,
    end: Coordinate,
    row_limit: usize,

    pub fn new(start: Coordinate, end: Coordinate, row_limit: usize) @This() {
        return .{ .current = start, .end = end, .row_limit = row_limit };
    }

    pub fn next(self: *@This()) ?Coordinate {
        if (self.current.cmp(.great, self.end)) {
            return null;
        }

        const output = self.current;

        if (self.current.x >= self.row_limit) {
            self.current.x = 0;
            self.current.y += 1;
        } else {
            self.current.x += 1;
        }

        return output;
    }
};
